{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protein Modification"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Find this notebook at `EpyNN/nnlive/ptm_protein/prepare_dataset.ipynb`. \n",
    "* Regular python code at `EpyNN/nnlive/ptm_protein/prepare_dataset.py`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is part of the series on preparing data for Neural Network regression with EpyNN. \n",
    "\n",
    "It deals with a real world problem and therefore will focus on the problem itself, rather than basics that were reviewed along with the preparation of the following dummy dataset: \n",
    "\n",
    "* [Boolean dataset](../dummy_boolean/prepare_dataset.ipynb)\n",
    "* [String dataset](../dummy_string/prepare_dataset.ipynb)\n",
    "* [Time-series (numerical)](../dummy_time/prepare_dataset.ipynb)\n",
    "* [Image (numerical)](../dummy_image/prepare_dataset.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Post Translational Modification (PTM) of Proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Post-Translational Modification (PTM) of proteins is an ensemble of mechanisms by which the primary sequence of a protein can be chemically modified after - and in some circumstances during - biosynthesis by the ribosomes.\n",
    "\n",
    "When talking about *one* PTM, it generally refers to a given chemical group that may be covalently linked with given amino acid residues in proteins.\n",
    "\n",
    "For instance, the formation of a phosphoester between a phosphate group and side-chain hydroxyl of serine, threonine and tyrosine is known as phosphorylation. While proteins overall may contain a given number of such residues, phosphorylation may occur particularly on a given subset, generally with respect to specific cellular conditions.\n",
    "\n",
    "From a given number of chemically unmodified proteins (proteome), below is a list of some characteristics with respect to PTM:\n",
    "\n",
    "* PTM increase chemical diversity: for a given *proteome*, there is a corresponding *phosphoproteome* or *oglcnacome* if talking about *O*-GlcNAcylation. Said explicitely, a chemically uniform protein may give rise to an ensemble of chemically distinct proteins upon modification.\n",
    "* PTM may enrich gene's function: as for other mechanisms, the fact that a given gene product - the chemically unmodified protein - may be modified to yield distinct chemical entities is equivalent to multiplying the number of end-products from a single gene. As such, the number of functions for this gene is expected to increase, because distinct functions are achieved by distinct molecules, and this is actually what PTM do: create chemically distinct proteins from the same gene product.\n",
    "* Chemical groups defining one PTM are numerous: among the most studied, one may cite phosphorylation, ubiquitinylation, *O*-GlcNActylation, methylation, succinylation, among dozens of others.\n",
    "\n",
    "PTMs are major regulators of cell signaling and play a role in virtually every biological process.\n",
    "\n",
    "As such, this is a big challenge to predict whether or not one protein may be modified with respect to one PTM.\n",
    "\n",
    "Let's draw something to illustrate one aspect of the deal."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![2hr9](tctp.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a protein called [TCTP](https://www.rcsb.org/structure/2hr9) and above is shown a type of 3D model commonly used to represent proteins. The red sticks represent serine residues along the protein primary sequence. Those with label SER-46 and SER-64 where shown to undergo phosphorylation in cells.\n",
    "\n",
    "But in theory, phosphorylation could occur on all serines within this structure. The reality is that such modifications only occur on *some* serines.\n",
    "\n",
    "This is what we are going to challenge here, with a PTM called *O*-GlcNAcylation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare a set of peptides"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let’s prepare a set of O-GlcNAcylated and presumably not *O*-GlcNAcylated peptide sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EpyNN/nnlive/ptm_protein/prepare_dataset.ipynb\n",
    "# Standard library imports\n",
    "import tarfile\n",
    "import random\n",
    "import os\n",
    "\n",
    "# Related third party imports\n",
    "import wget\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Local application/library specific imports\n",
    "from epynn.commons.library import read_file\n",
    "from epynn.commons.logs import process_logs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note the `tarfile` which is a Python built-in *standard* library and the first choice to deal with `.tar` archives and related."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Seeding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For reproducibility."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simple function to download data from the cloud as `.tar` archive. Once uncompressed, it yields a `data/` directory containing `.dat` text files for positive and negative sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_sequences():\n",
    "    \"\"\"Download a set of peptide sequences.\n",
    "    \"\"\"\n",
    "    data_path = os.path.join('.', 'data')\n",
    "\n",
    "    if not os.path.exists(data_path):\n",
    "\n",
    "        # Download @url with wget\n",
    "        url = 'https://synthase.s3.us-west-2.amazonaws.com/ptm_prediction_data.tar'\n",
    "        fname = wget.download(url)\n",
    "\n",
    "        # Extract archive\n",
    "        tar = tarfile.open(fname).extractall('.')\n",
    "        process_logs('Make: ' + fname, level=1)\n",
    "\n",
    "        # Clean-up\n",
    "        os.remove(fname)\n",
    "\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retrieve the data as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "download_sequences()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('data', [], ['21_positive.dat', '21_negative.dat'])\n"
     ]
    }
   ],
   "source": [
    "for path in os.walk('data'):\n",
    "    print(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let’s have a quick look to what one file's content."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SQDVSNAFSPSISKAQPGAPP', 'GPRIPDHQRTSVPENHAQSRI', 'QFSCKCLTGFTGQKCETDVNE', 'KLIKRLYVDKSLNLSTEFISS', 'QQKEGEQNQQTQQQQILIQPQ']\n"
     ]
    }
   ],
   "source": [
    "with open(os.path.join('data', '21_positive.dat'), 'r') as f:\n",
    "    print(f.read().splitlines()[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These are 21 amino-acids long peptide sequences.\n",
    "\n",
    "Note that positive sequences are *Homo sapiens* *O*-GlcNAcylated peptides sourced from [The *O*-GlcNAc Database](https://www.oglcnac.mcw.edu).\n",
    "\n",
    "Negative sequences are *Homo sapiens* peptide sequence not reported in the above-mentioned source. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is a function we use to prepare the labeled dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_dataset(N_SAMPLES=100):\n",
    "    \"\"\"Prepare a set of labeled peptides.\n",
    "\n",
    "    :param N_SAMPLES: Number of peptide samples to retrieve, defaults to 100.\n",
    "    :type N_SAMPLES: int\n",
    "\n",
    "    :return: Set of peptides.\n",
    "    :rtype: tuple[list[str]]\n",
    "\n",
    "    :return: Set of single-digit peptides label.\n",
    "    :rtype: tuple[int]\n",
    "    \"\"\"\n",
    "    # Single-digit positive and negative labels\n",
    "    p_label = 0\n",
    "    n_label = 1\n",
    "\n",
    "    # Positive data are Homo sapiens O-GlcNAcylated peptide sequences from oglcnac.mcw.edu\n",
    "    path_positive = os.path.join('data', '21_positive.dat')\n",
    "\n",
    "    # Negative data are peptide sequences presumably not O-GlcNAcylated\n",
    "    path_negative = os.path.join('data', '21_negative.dat')\n",
    "\n",
    "    # Read text files, each containing one sequence per line\n",
    "    positive = [[list(x), p_label] for x in read_file(path_positive).splitlines()]\n",
    "    negative = [[list(x), n_label] for x in read_file(path_negative).splitlines()]\n",
    "\n",
    "    # Shuffle data to prevent from any sorting previously applied\n",
    "    random.shuffle(positive)\n",
    "    random.shuffle(negative)\n",
    "\n",
    "    # Truncate to prepare a balanced dataset\n",
    "    negative = negative[:len(positive)]\n",
    "\n",
    "    # Prepare a balanced dataset\n",
    "    dataset = positive + negative\n",
    "\n",
    "    # Shuffle dataset\n",
    "    random.shuffle(dataset)\n",
    "\n",
    "    # Truncate dataset to N_SAMPLES\n",
    "    dataset = dataset[:N_SAMPLES]\n",
    "\n",
    "    # Separate X-Y pairs\n",
    "    X_features, Y_label = zip(*dataset)\n",
    "\n",
    "    return X_features, Y_label"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check the function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 ['T', 'A', 'A', 'M', 'R', 'N', 'T', 'K', 'R', 'G', 'S', 'W', 'Y', 'I', 'E', 'A', 'L', 'A', 'Q', 'V', 'F']\n",
      "1 ['N', 'K', 'K', 'L', 'A', 'P', 'S', 'S', 'T', 'P', 'S', 'N', 'I', 'A', 'P', 'S', 'D', 'V', 'V', 'S', 'N']\n",
      "1 ['R', 'G', 'A', 'G', 'S', 'S', 'A', 'F', 'S', 'Q', 'S', 'S', 'G', 'T', 'L', 'A', 'S', 'N', 'P', 'A', 'T']\n",
      "0 ['T', 'D', 'N', 'D', 'W', 'P', 'I', 'Y', 'V', 'E', 'S', 'G', 'E', 'E', 'N', 'D', 'P', 'A', 'G', 'D', 'D']\n",
      "0 ['G', 'Q', 'E', 'R', 'F', 'R', 'S', 'I', 'T', 'Q', 'S', 'Y', 'Y', 'R', 'S', 'A', 'N', 'A', 'L', 'I', 'L']\n",
      "0 ['S', 'I', 'N', 'T', 'G', 'C', 'L', 'N', 'A', 'C', 'T', 'Y', 'C', 'K', 'T', 'K', 'H', 'A', 'R', 'G', 'N']\n",
      "1 ['N', 'K', 'A', 'S', 'L', 'P', 'P', 'K', 'P', 'G', 'T', 'M', 'A', 'A', 'G', 'G', 'G', 'G', 'P', 'A', 'P']\n",
      "1 ['A', 'S', 'V', 'Q', 'D', 'Q', 'T', 'T', 'V', 'R', 'T', 'V', 'A', 'S', 'A', 'T', 'T', 'A', 'I', 'E', 'I']\n",
      "1 ['A', 'S', 'L', 'E', 'G', 'K', 'K', 'I', 'K', 'D', 'S', 'T', 'A', 'A', 'S', 'R', 'A', 'T', 'T', 'L', 'S']\n",
      "1 ['R', 'R', 'Q', 'P', 'V', 'G', 'G', 'L', 'G', 'L', 'S', 'I', 'K', 'G', 'G', 'S', 'E', 'H', 'N', 'V', 'P']\n"
     ]
    }
   ],
   "source": [
    "X_features, Y_label = prepare_dataset(N_SAMPLES=10)\n",
    "\n",
    "for peptide, label in zip(X_features, Y_label):\n",
    "    print(label, peptide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These sequences are centered with respect to the modified or presumably unmodified residue, which may be a serine or a threonine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 ['S']\n",
      "1 ['S']\n",
      "1 ['S']\n",
      "0 ['S']\n",
      "0 ['S']\n",
      "0 ['T']\n",
      "1 ['T']\n",
      "1 ['T']\n",
      "1 ['S']\n",
      "1 ['S']\n"
     ]
    }
   ],
   "source": [
    "for peptide, label in zip(X_features, Y_label):\n",
    "    print(label, peptide[len(peptide) // 2:len(peptide) // 2 + 1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because *O*-GlcNAcylation may impact Serine or Threonine, note that negative sequences with label ``0`` were prepared to also contain such residues at the same position."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have already seen in [String dataset](../dummy_string/prepare_dataset.ipynb) how to perform [*one-hot encoding*](../dummy_string/prepare_dataset.ipynb#One-hot-encoding-of-string-features) of string features.\n",
    "\n",
    "Just for fun, and also because you may like to use such data in convolutional networks, let's convert a peptide sequence into an image."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n",
      "{'T': 0, 'S': 1, 'E': 2, 'L': 3, 'M': 4, 'A': 5, 'W': 6, 'C': 7, 'I': 8, 'P': 9, 'D': 10, 'V': 11, 'Q': 12, 'H': 13, 'R': 14, 'Y': 15, 'G': 16, 'K': 17, 'N': 18, 'F': 19}\n",
      "['G', 'R', 'I', 'S', 'A', 'L', 'Q', 'G', 'K', 'L', 'S', 'K', 'L', 'D', 'Y', 'R', 'D', 'I', 'T', 'K', 'Q']\n",
      "[16, 14, 8, 1, 5, 3, 12, 16, 17, 3, 1, 17, 3, 10, 15, 14, 10, 8, 0, 17, 12]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAD4AAAD4CAYAAAC0cFXtAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAALdElEQVR4nO2dfYwU9RnHP18PadGDEzDiGwdaCAkxSnuE2ogGX4to1L5DGoutFutLosamsTXRxqYJbaO0DQbry0VtRGl8SYklIlWpJamUk+ALAuW0egWpKFwPTimXO5/+Mb+lyzJ7Ozezezvc7/dJNjv7m2fmN9+d2ZnZ55nn+cnM8JEj6r0B9SII940g3DeG1XsD4hg7dqw1NzdXtOvo6GDXrl1K00cuhTc3N7N69eqKdrNmzUrdR6ZDXdJsSVsktUu6LWb+ZyQtc/PXSpqYpb9qklq4pAbgXuBiYCowT9LUErOrgU4zmwQsAn6Rtr9qk2WPzwDazewdM+sBngAuL7G5HHjETT8JnC8p1W+y2mQRfhLwr6LP21xbrI2Z9QJdwNgMfVaN3FzOJC2Q1CapbdeuXTXvL4vw7cD4os8nu7ZYG0nDgCYgVpWZ3W9m081s+tixtT8osghfB0yWdIqk4cBcYHmJzXJgvpv+OvCi5eTvYOrruJn1SroRWAk0AK1mtlHSXUCbmS0HHgJ+L6kd2E305eSCTDcwZrYCWFHSdkfR9H+Bbwx0vT09PXR0dCSyS0tuTm6DTRDuG0G4bwThvhGE+4a3wnPpc9u9ezdLly5NZJcWb/d4EO4bQbhvBOEDRdJ4SS9JekvSRkk3xdjMktQlaYN73RG3rnqQ5TreC9xqZusljQRelbTKzN4qsfurmV2aoZ+akHqPm9kOM1vvpvcCmzg0oJBbVA1vrwsGvgycZmZ7itpnAU8RRVneB35oZhvLrGMBsMBNt4waNapiv93d3fT29qYKSWUWLqkR+AvwczN7umTeKOBTM+uWNAf4jZlNrrTOYcOGWWNjY8W+swjPGiY+kmiPPlYqGsDM9phZt5teARwp6dgsfVaLLGd1EQUMNpnZPWVsji9ERyXNcP3VPjCWgCxn9bOAK4E3JG1wbT8BmgHM7D6isNF1knqBfcDcvISQqnJyqza5/40fzgThvhGE+0YunY2NjY3MnDmzot2aNWtS9+HtHg/CfSMI940g3DeCcN/I5Z3b/v37ee+99xLZpcXbPZ5ZuKR3Jb3hIiVtMfMl6bcuL+V1SV/I2mc1qNahfq6ZfVRm3sXAZPf6IrDEvdeVwTjULwcetYhXgGMknTAI/fZLNYQb8LykV100pJQkuSsHpWb09fVVYbP6pxqH+kwz2y7pOGCVpM1m9vJAV2Jm9wP3A4wYMaLmrt/Me9zMtrv3ncAzRGlZxSTJXRl0soaQjnYhYiQdDVwEvFlithz4jju7nwl0mdmOLP1Wg6yH+jjgGRclGgYsNbPnJP0ADkRTVgBzgHbgE+C7GfusCrmMpEyePNkWLVpU0e6WW25h69atIZIyEIJw3wjCfSMI940g3De8FZ5LZ2NjYyNnn312Iru0eLvHg3DfCMJ9IwgfKJKmFOWabJC0R9LNJTZDLyfFzLYA0+BApa/tRF7WUoZWTkoJ5wNvm1nlEGdOqNad21zg8TLzviTpNQaQk9Lc3ExTU1PFThsaGtJtLYCZZXoBw4GPgHEx80YBjW56DrA1yTpbWlosCc4u1XZX41C/GFhvZh/EfKlDLyeliHmUOcyHak5KIWx0IXBtUVtxFCW3OSlZq3t9TEkpQie4ML0YWJylj1oR7tx8Iwj3jVyGiUeMGGGTJk2qaNfe3s6+fftCmHggBOG+EYT7RhDuG0G4b3grPJe3rKFGRA1JJFxSq6Sdkt4sahsjaZWkre59dJll5zubrZLmx9nUg6R7/GFgdknbbcALroTRC+7zQUgaA9xJlIMyA7iz3Bc02CQS7jIOSmuBFo+B8ghwRcyiXwZWmdluM+sEVnHoF1gXsjgbxxU9cP9vomfXS0mUjwKHVPfKsFnJqMrJzbmMM10erGi4kCOOqP05N0sPHxTSqNz7zhibXOajQDbhxWOgzAf+GGOzErhI0mh3UrvItdWdpJezx4G/AVMkbZN0NbAQuFDSVuAC9xlJ0yU9CGBmu4GfEQ0msw64y7XVn6zR0lq8Dpdo6WFJEO4bQbhvBOG+EYT7hrfCc+lsDPHxGhKE+0YQ7htBeDnKRFF+JWmzq9b1jKRjyizbb+WvepJkjz/MoUGAVUQjZJwO/AP4cT/Ln2tm08xserpNrA0VhcdFUczseYsGUgd4hchtfFhRjZyU7wHLyswrVP4y4HcWFbKKpTiSMmbMGK6//vqKHS9cuHDgW+vI+qD+7URDAz1WxiRx5S8rqu41YcKE/Fb3knQVcCnw7XJZB1a58lfdSCVc0mzgR8BlZvZJGZsklb/qRpLLWVwUZTEwkujw3SDpPmd7oqTCQOzjgDUu5+zvwJ/M7LmaqEhBxd+4mc2LaX6ojO37RPllmNk7wBmZtq6GhDs33wjCfSOXdWCampq45JJLKtotWbIkdR/e7vEg3DeCcN8Iwn0jCPcNb4Xn8pa1r6+Prq6uRHZp8XaPp42k/FTS9qKqXXPKLDtb0hY3RsohqRv1JG0kBWCRi5BMs6iqz0G4il/3ElUGmgrMkzQ1y8ZWk1SRlITMANrN7B0z6wGeIMpjyQVZfuM3uqBha5nMosT5KHDwOCmdnZ0ZNisZaYUvAT5HVMhuB3B31g0pzkkZPbr2GVqphJvZB2bWZ2afAg8QHyHJbT4KpI+kFI9l9BXiIyTrgMmSTpE0nKjY3fI0/dWCijcwLpIyCzhW0jaizMFZkqYRRUPfxVX3knQi8KCZzTGzXkk3EiXfNACtVqZ6Xz2oWSTFfV5BNEDMgOju7k407nB3d/dAV32AcOfmG0G4bwThvhGE+4a3wnPpc+vp6aGjoyORXVq83eNBuG8E4b4RhPtGEtdTK9Hj2TvN7DTXtgyY4kyOAf5jZtNiln0X2Av0Ab15Ss9IcgPzMNHTyo8WGszsW4VpSXcD/UX4+huNvm4k8bm9LGli3Dw3PsI3gfOqvF21J0mxGGAi8GZM+zlAWz/L/RNYD7wKLKjQxwKgzb0s6SttAZyswpcAt/az3Enu/TjgNeCchP3VXHiWnJRhwFcpn4E09HJSHBcAm81sW9zMoZqTAjHjHx1OOSm5LJXgEvQSYWahVMJACMJ9IwjPEy0tLUlLHKbuI5fCB4Mg3DeCcN8Iwn0jCPcNb4XnMj7e1dXFs88+m8guLUk8MOMlvSTpLUkbJd3k2of8cCG9RJ7UqcCZwA0uxWJoDxdiZjvMbL2b3gtsIso0OKyHCxnQyc1FVD4PrKXKw4UMNomFS2oEngJuNrM9xfMs8lhm8loW56RkOWklJengEUcSiX7MzJ52zVUdLsSKclKampqSbn9qkpzVRfRg/iYzu6do1pAfLuQs4ErgvJLMwsN6uJAkYeI1QDmn/fkx9m3ANUWfW4HWtBtYK/IaSdkLbClpPhYofcBgipmNTNNHLm9ZgS2lj41IaotrS9uBt39SgvCcEVfGNGlbInJ5chsM8rrHa04QXk9KnBrr3fuBuhJFtSZM0g5JnZI+lrRW0kRJV0n6sOjO8ppKfaZ6VKraL+CXRI6MBmAXcB8wnOgRsdOAt4FTgW6iv7bL3HJziZ66ugpYPCiPe1WZglNjBvA60WOghboSN+BqTTjbT/h/zYonibltTkJehBecGicR7d2CU2Mb0cOFBWfGZ93nr0m6wqJS511Ao2t7XdKTkor/CscyaLeskv4MHB8z6/aYtnLX2AlE9dzXAr+W9IZrXwk8YGb7JV1LdPT0+3ztoAk3swvKzZNUcGpsJyq6UXBqnExUkeBUt47tkvYBHwKrgRagieinUPiyHiQ6Z/RLXg71glNjHXA6sLqorsQSoloTZzj311HACUR+gvHAixx8JF1G5BDtl1zcuUkaC/wBaCY6c48k8gGsJDq7HwVcCIwCOoERRFeA94m8tt8nEtxLdOK7zsw299tnHoTXg7wc6oNOEO4bQbhvBOG+8T/QzncT1bt/4AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "X_features, _ = prepare_dataset(N_SAMPLES=10)\n",
    "\n",
    "# Flatten the list of lists (list of peptides) and make a set()\n",
    "aas = list(set([feature for features in X_features for feature in features]))\n",
    "\n",
    "# set() contains unique elements\n",
    "print(len(aas))  # 20 amino-acids\n",
    "\n",
    "e2i = {k: i for i, k in enumerate(aas)}  # element_to_idx encoder 0-19\n",
    "\n",
    "features = X_features[0]\n",
    "\n",
    "print(e2i)       # Encoder\n",
    "print(features)  # Peptide before encoding\n",
    "print([e2i[feature] for feature in features])  # After encoding\n",
    "\n",
    "# NumPy array to plot as image\n",
    "img_features = np.array([e2i[feature] for feature in features])\n",
    "img_features = np.expand_dims(img_features, axis=1)\n",
    "\n",
    "plt.imshow(img_features, cmap='gray')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well, let’s reshape. The number 21 is divisible by 7 and 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAHgAAAD4CAYAAAAn1CIKAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAIkElEQVR4nO3dXYhV1xnG8f/TMdIPgwm2lBC1UQiCFNoEEUo0UEuLSUtz04sEGmgJeNOURAolXvbKu9heFEESa6FOJOQDgqRJA40EIU2TWFP8SOpUqo6kaPyKqdJB+/Zi9oSJTjNrZtaac+bt84Mh52yPe63hnz2zPeessxURWF6f6fUErC0HTs6Bk3Pg5Bw4uXktdrpo0aJYunRpi11PamRkpCfjjjl79mxPxr1w4QKXLl3StdubBF66dCl79uxpsetJHT9+vCfjjhkcHOzJuDt27Jhwu39EJ+fAyTlwcg6cnAMn58DJOXByDpycAyfnwMk5cHJFgSWtl/SepCFJj7WelNUzaWBJA8CvgXuAlcADkla2npjVUXIErwaGIuJoRIwAu4D72k7LaikJfCtwYtz94W7bJ0jaIOktSW+dOXOm1vxshqqdZEXEtohYFRGrFi1aVGu3NkMlgU8CS8bdX9xtszmgJPCbwO2SlkmaD9wPvNB2WlbLpG/ZiYgrkh4GXgYGgO0RcbD5zKyKovdkRcSLwIuN52IN+Jms5Bw4OQdOzoGTc+DkHDg5B07OgZNz4OTU4kNY5s2bFwsWLKi+3xJr1qzpybhjjh071pNxh4aGuHz58nXLR30EJ+fAyTlwcg6cnAMn58DJOXByDpycAyfnwMk5cHIOnFzJ6sLtkk5JOjAbE7K6So7gHcD6xvOwRiYNHBGvAb35CFWbsWqfNitpA7Chu11rtzZD1QJHxDZgG4y+4F9rvzYzPotOzoGTK/ln0lPA68AKScOSHmo/LaulZH3wA7MxEWvDP6KTc+DkHDg5B07OgZNz4OQcODkHTs6Bk2tyccply5axZcuWFrue1Nq1a3sy7piFCxf2ZNxVq1ZNuN1HcHIOnJwDJ+fAyTlwcg6cnAMn58DJOXByDpycAyfnwMmVvC96iaRXJR2SdFDSI7MxMauj5NWkK8DPImKfpBuBtyW9EhGHGs/NKihZPvp+ROzrbl8EDjPBxSmtP03pd7Ck24A7gDcm+LOPrz564cKFStOzmSoOLGkB8CzwaER8eO2fj7/6aK9e9LbrlV7i/QZG4+6MiOfaTslqKjmLFvAkcDgiHm8/Jaup5Ai+C3gQWCdpf/d1b+N5WSUly0f3Av7QjTnKz2Ql58DJOXByDpycAyfnwMk5cHIOnJwDJ9dk+ejw8DCbNm1qsetJnThxoifjjjl//nxPx7+Wj+DkHDg5B07OgZNz4OQcODkHTs6Bk3Pg5Bw4OQdOzoGTK3nj+2cl/VnSO93y0V/MxsSsjpJXk/4NrIuIj7olLHsl/T4i/tR4blZByRvfA/iou3tD9+VrE84RpYvPBiTtB04Br0TEpy4fvXr1auVp2nQVBY6IqxHxdWAxsFrSVyd4zMfLRwcGBipP06ZrSmfREXEeeBVfEXzOKDmL/pKkm7rbnwO+DbzbeF5WSclZ9C3AbyUNMPo/xNMRsbvttKyWkrPovzL6uRw2B/mZrOQcODkHTs6Bk3Pg5Bw4OQdOzoGTc+DkHDi5JuuDe2nz5s09Hf/48eM9GXdkZGTC7T6Ck3Pg5Bw4OQdOzoGTc+DkHDg5B07OgZNz4OQcOLmpXPlsQNJfJPk90XPIVI7gRxi9MKXNIaWrCxcD3wWeaDsdq630CP4l8HPgP//rAV4+2p9KFp99DzgVEW9/2uO8fLQ/lV678PuS/gHsYvQahr9rOiurpuQK4JsiYnFE3AbcD/wxIn7YfGZWhf8dnNyU3pMVEXuAPU1mYk34CE7OgZNz4OQcODkHTs6Bk3Pg5Bw4OQdOzoGTa7J8dPny5QwODrbY9aT27t3bk3HHbN26tSfjnj59esLtPoKTc+DkHDg5B07OgZNz4OQcODkHTs6Bk3Pg5Bw4uaLnortVDReBq8CViFjVclJWz1RebPhmRHzQbCbWhH9EJ1caOIA/SHpb0oaJHjB++ei5c+fqzdBmpDTwmoi4E7gH+Imku699wPjlozfffHPVSdr0lV5e9mT331PA88DqlpOyekoWgH9B0o1jt4HvAAdaT8zqKDmL/jLwvKSxxw9GxEtNZ2XVlFx99CjwtVmYizXgfyYl58DJOXByDpycAyfnwMk5cHIOnJwDJ+fAySki6u9Uqr/TQi2+n6nYvbs3H4i/ceNGjhw5omu3+whOzoGTc+DkHDg5B07OgZNz4OQcODkHTs6Bk3Pg5EqvXXiTpGckvSvpsKRvtJ6Y1VG6fPRXwEsR8QNJ84HPN5yTVTRpYEkLgbuBHwFExAgw0nZaVkvJj+hlwGngN90Fop/o1ih9wvjlo9VnadNWEngecCewNSLuAP4FPHbtg8YvH608R5uBksDDwHBEvNHdf4bR4DYHlFx99J/ACUkruk3fAg41nZVVU3oW/VNgZ3cGfRT4cbspWU1FgSNiP+DfrXOQn8lKzoGTc+DkHDg5B07OgZNz4OQcODkHTs6Bk2ty9VHgA+DYNP/uF7u/Py3dRy5O14zGnqGZjv2ViTY2WR88E5Le6tVryhnH9o/o5Bw4uX4MvM1j19N3v4Otrn48gq0iB06urwJLWi/pPUlDkq57a27DcbdLOiVp1q9FIWmJpFclHZJ0UNIjVQeIiL74AgaAvwPLgfnAO8DKWRr7bkbfCnygB9/3LcCd3e0bgb/V/L776QheDQxFxNFuecwu4L7ZGDgiXgPOzsZYE4z9fkTs625fBA4Dt9bafz8FvhU4Me7+MBW/0blA0m3AHcAbkzy0WD8F/r8maQHwLPBoRHxYa7/9FPgksGTc/cXdtvQk3cBo3J0R8VzNffdT4DeB2yUt61ZQ3A+80OM5NafRl7+eBA5HxOO19983gSPiCvAw8DKjJxpPR8TB2Rhb0lPA68AKScOSHpqNcTt3AQ8C6yTt777urbVzP1WZXN8cwdaGAyfnwMk5cHIOnJwDJ+fAyf0XVNhXGsblI90AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.imshow(img_features.reshape(7, 3), cmap='gray')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems to be working! We are done."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Live examples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function ``prepare_dataset()`` presented herein is used in the following live examples:\n",
    "\n",
    "* Notebook at`EpyNN/nnlive/dummy_string/train.ipynb` or following [this link](train.ipynb). \n",
    "* Regular python code at `EpyNN/nnlive/dummy_string/train.py`."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
